# 1. Load and Filter Benchmark Results

- **Load benchmark results** from the CSV file `benchmark_results.csv`.
- **Filter out rows** where either `Runtime (s)` (measured runtime) or `Reported Runtime (s)` is missing. This ensures only runs with valid runtime data are considered.
- **Keep only successful solver runs** where `Status == "ok"`, ignoring failed or timed-out runs.

In [19]:
import sys

import pandas as pd

sys.path.insert(0, "..")
from runner.utils import load_benchmark_metadata

# Load benchmark results into benchmark_data
benchmark_data = pd.read_csv("../results/benchmark_results.csv")

# Keep only rows where both measured and reported runtimes exist
benchmark_data = benchmark_data[
    (~benchmark_data["Runtime (s)"].isna())
    & (~benchmark_data["Reported Runtime (s)"].isna())
]

print(f"Rows with both runtimes available: {len(benchmark_data)}")

# Keep only successful solver runs
benchmark_data = benchmark_data[benchmark_data["Status"] == "ok"]

print(f"Rows after filtering Status==ok: {len(benchmark_data)}")

Rows with both runtimes available: 2528
Rows after filtering Status==ok: 1615


# 2. Compute Runtime Differences and Add Identifier Columns

- **Compute absolute runtime difference** (`runtime-difference`) between the measured runtime (`Runtime (s)`) and the reported runtime (`Reported Runtime (s)`).
- **Compute relative runtime difference** (`runtime-difference-%`) as a percentage of the larger of the two runtimes. 
- **Construct `bench-size` column** as a unique identifier for each benchmark instance by combining the benchmark name and size.
- **Construct `solver-version` column** as a unique identifier for each solver run by combining the solver name and its version.

In [20]:
# Compute absolute and percentage runtime differences
benchmark_data["runtime-difference"] = (
    benchmark_data["Runtime (s)"] - benchmark_data["Reported Runtime (s)"]
).abs()

benchmark_data["runtime-difference-%"] = (
    benchmark_data["runtime-difference"]
    / benchmark_data[["Runtime (s)", "Reported Runtime (s)"]].max(axis=1)
) * 100

# Construct bench-size and solver-version in benchmark_data
benchmark_data["bench-size"] = (
    benchmark_data["Benchmark"] + "-" + benchmark_data["Size"]
)

benchmark_data["solver-version"] = (
    benchmark_data["Solver"] + "-" + benchmark_data["Solver Version"]
)

# Keep only relevant columns for further analysis
cols_to_export = [
    "bench-size",
    "solver-version",
    "Runtime (s)",
    "Reported Runtime (s)",
    "runtime-difference",
    "runtime-difference-%",
]

benchmark_data = benchmark_data[cols_to_export]

# Display the first few rows of `benchmark_data`
display(benchmark_data.head())

,bench-size,solver-version,Runtime (s),Reported Runtime (s),runtime-difference,runtime-difference-%
0,TIMES-GEO-global-netzero-31-20ts,gurobi-13.0.0,30621.769838,30590.708018,31.061819,0.101437
6,pypsa-de-elec-50-1h,gurobi-13.0.0,2989.140083,2923.996410,65.143673,2.179345
12,temoa-US_9R_TS-9-12ts,gurobi-13.0.0,190.656273,172.031106,18.625167,9.768977
13,temoa-US_9R_TS-9-12ts,highs-hipo-1.12.0-hipo,3181.154351,3181.154351,0.000000,0.000000
14,temoa-US_9R_TS-9-12ts,highs-ipx-1.12.0-hipo,3388.801879,3388.801879,0.000000,0.000000


# 3. Load and Prepare Benchmark Metadata

- Load metadata from `metadata.yaml` using `load_benchmark_metadata()`. This contains details about each benchmark instance, such as number of variables, number of constraints, and size category.

- Construct `bench-size` column as a unique identifier for each benchmark instance by combining `Benchmark` and `Instance`. This will be used to merge metadata with runtime results.

- Copy `Size` column to a new column called `bench-size-category` to keep the size category information (S, M, L, etc.) separate.

In [21]:
# Load benchmark metadata
benchmark_metadata = load_benchmark_metadata("../results/metadata.yaml")

# Construct bench-size in benchmark_metadata
benchmark_metadata["bench-size"] = (
    benchmark_metadata["Benchmark"] + "-" + benchmark_metadata["Instance"]
)

# Copy Size column to a new column called "bench-size-category"
benchmark_metadata["bench-size-category"] = benchmark_metadata["Size"]

# Keep only relevant metadata fields
meta_cols = [
    "bench-size",
    "Num. variables",
    "Num. constraints",
    "bench-size-category",
]

benchmark_metadata = benchmark_metadata[meta_cols]

# Display the first few rows of `benchmark_metadata`
display(benchmark_metadata.head())

,bench-size,Num. variables,Num. constraints,bench-size-category
ethos_fine_europe_60tp-175-720ts,ethos_fine_europe_60tp-175-720ts,8141182,9230938,L
ethos-fine-multi-regional-7tp-8-168ts,ethos-fine-multi-regional-7tp-8-168ts,55979,91178,M
ethos-fine-multi-regional-7tp-12seg-8-84ts,ethos-fine-multi-regional-7tp-12seg-8-84ts,33215,64214,M
ethos-fine-energyland-full-timeseries-1-8760ts,ethos-fine-energyland-full-timeseries-1-8760ts,473179,543229,M
ethos-fine-energyland-48tp-1-1152ts,ethos-fine-energyland-48tp-1-1152ts,32770,40198,M


# 4. Merge Benchmark Results with Metadata and Export

- **Merge `benchmark_data` with `benchmark_metadata`** on the `bench-size` column to combine runtime results with the relevant metadata.

- **Save the merged DataFrame** to a CSV file (`runtime_reported_runtime_discrepancies.csv`) for later analysis.  

In [22]:
# Merge benchmark_data with benchmark_metadata
comparison_df = benchmark_data.merge(
    benchmark_metadata,
    on="bench-size",
    how="left",
)

# Save merged data to CSV
output_csv = "../results/runtime_reported_runtime_discrepancies.csv"
comparison_df.to_csv(output_csv, index=False)

# Display the first few rows of `comparison_df`
display(comparison_df.head())

,bench-size,solver-version,Runtime (s),Reported Runtime (s),runtime-difference,runtime-difference-%,Num. variables,Num. constraints,bench-size-category
0,TIMES-GEO-global-netzero-31-20ts,gurobi-13.0.0,30621.769838,30590.708018,31.061819,0.101437,5422595,5063755,L
1,pypsa-de-elec-50-1h,gurobi-13.0.0,2989.140083,2923.996410,65.143673,2.179345,8637885,18370417,L
2,temoa-US_9R_TS-9-12ts,gurobi-13.0.0,190.656273,172.031106,18.625167,9.768977,1801794,1594079,L
3,temoa-US_9R_TS-9-12ts,highs-hipo-1.12.0-hipo,3181.154351,3181.154351,0.000000,0.000000,1801794,1594079,L
4,temoa-US_9R_TS-9-12ts,highs-ipx-1.12.0-hipo,3388.801879,3388.801879,0.000000,0.000000,1801794,1594079,L


# 5. Filter and Sort Benchmark Results (Runtime ≥ 60s)

In [24]:
comparison_df_1min = comparison_df[comparison_df["Runtime (s)"] >= 60]

print(f"Rows after filtering Runtime >= 60s: {len(comparison_df_1min)}")

Rows after filtering Runtime >= 60s: 807


# 6. Identify Benchmark–Solver Pairs with Largest Runtime Discrepancies (Size S)

In [29]:
top_S = (
    comparison_df_1min[comparison_df_1min["bench-size-category"] == "S"]
    .sort_values("runtime-difference-%", ascending=False)
    .head(5)
)

print("Top 5 discrepancies — Small (S):")
display(top_S)

Top 5 discrepancies — Small (S):


,bench-size,solver-version,Runtime (s),Reported Runtime (s),runtime-difference,runtime-difference-%,Num. variables,Num. constraints,bench-size-category
1285,pglib_opf_case162_ieee_dtc-162-NA,gurobi-10.0.0,70.708702,70.388483,0.320219,0.452871,742,1867,S
1327,pglib_opf_case162_ieee_dtc-162-NA,gurobi-12.0.0,60.193940,59.923420,0.270520,0.449415,742,1867,S
1583,Sienna_modified_RTS_GMLC_DA_sys_NetPTDF_Horizo...,cbc-2.10.12,88.164783,87.880000,0.284783,0.323012,8820,10368,S
414,Sienna_modified_RTS_GMLC_DA_sys_NetTransport_H...,cbc-2.10.12,65.350960,65.220000,0.130960,0.200395,8820,9792,S
231,Sienna_modified_RTS_GMLC_DA_sys_NetPTDF_Horizo...,cbc-2.10.12,155.051203,154.760000,0.291203,0.187811,8820,10368,S


# 7. Identify Benchmark–Solver Pairs with Largest Runtime Discrepancies (Size M)

In [26]:
top_M = (
    comparison_df_1min[comparison_df_1min["bench-size-category"] == "M"]
    .sort_values("runtime-difference-%", ascending=False)
    .head(5)
)

print("Top 5 discrepancies — Medium (M):")
display(top_M)

Top 5 discrepancies — Medium (M):


,bench-size,solver-version,Runtime (s),Reported Runtime (s),runtime-difference,runtime-difference-%,Num. variables,Num. constraints,bench-size-category
764,tulipa-1_EU_investment_simple-28-2.2h,gurobi-10.0.0,63.276599,58.423052,4.853547,7.670367,745559,1118208,M
476,genx-2_three_zones_w_electrolyzer-no_uc-3-1h,gurobi-10.0.0,131.520907,123.789257,7.731650,5.878647,153413,175579,M
776,tulipa-1_EU_investment_simple-28-2.2h,gurobi-11.0.0,69.600160,65.792378,3.807782,5.470938,745559,1118208,M
789,tulipa-1_EU_investment_simple-28-2.2h,gurobi-12.0.0,98.688983,94.297260,4.391723,4.450064,745559,1118208,M
805,tulipa-1_EU_investment_simple-28-2.2h,gurobi-13.0.0,90.101743,86.207194,3.894549,4.322390,745559,1118208,M


# 8. Identify Benchmark–Solver Pairs with Largest Runtime Discrepancies (Size L)

In [27]:
top_L = (
    comparison_df_1min[comparison_df_1min["bench-size-category"] == "L"]
    .sort_values("runtime-difference-%", ascending=False)
    .head(5)
)

print("Top 5 discrepancies — Large (L):")
display(top_L)

Top 5 discrepancies — Large (L):


,bench-size,solver-version,Runtime (s),Reported Runtime (s),runtime-difference,runtime-difference-%,Num. variables,Num. constraints,bench-size-category
811,SWITCH-China-open-model-32-433ts,gurobi-13.0.0,716.099567,593.797664,122.301903,17.078896,13338520,14588407,L
1211,pypsa-de-elec-10-1h,gurobi-13.0.0,146.859648,128.301935,18.557713,12.636359,2111271,4502789,L
2,temoa-US_9R_TS-9-12ts,gurobi-13.0.0,190.656273,172.031106,18.625167,9.768977,1801794,1594079,L
1224,pypsa-de-elec-50-3h,gurobi-13.0.0,289.893353,263.912131,25.981222,8.962338,2879645,6123937,L
12,temoa-US_9R_TS_NZ-9-12ts,gurobi-13.0.0,225.330827,205.231677,20.099150,8.919840,1801794,1594085,L


## Conclusion

- **Small (S) Benchmarks *(Runtime ≥ 60s)***  
  - Absolute runtime differences are very small (≈0.1–0.3 seconds).  
  - Relative differences are also low (typically < 0.5%).  
  - Likely causes include minor measurement variations and solver initialization overhead.

- **Medium (M) Benchmarks *(Runtime ≥ 60s)***  
  - Moderate runtime differences (~4–8%), with absolute differences around ~3–8 second.  
  - Likely causes include LP/MPS file parsing, data handling, and preprocessing outside the solver’s internal timing.

- **Large (L) Benchmarks *(Runtime ≥ 60s)***  
  - Large absolute differences (tens to over a hundred seconds), while percentage differences remain moderate (~9–17%).  
  - Likely causes include parsing of large LP/MPS files, solver initialization overhead, license validation (especially for Gurobi), and additional setup not included in solver-reported runtime.

- **General Observations**  
  - Benchmark instances with runtime < 60 seconds were excluded to avoid misleadingly large percentage discrepancies due to very small runtimes.  
  - No systematic inconsistency between measured and solver-reported runtimes is observed across benchmarks or solvers.  
  - Differences are primarily due to overhead outside the solver’s internal timing, reflecting the distinction between full wall-clock execution and solver-reported computation time.